File: packages/sdks/maximem-synap/maximem_synap/cache/short_term_store.py
"""In-memory short-term context store, keyed by conversation_id.

The SDK becomes the authoritative live store of per-conversation
short-term (ST) context once the ``sdk_st_authoritative`` feature flag
is on for an instance.

Lifecycle is event-driven, not TTL-driven:
  * A ``compaction_update`` bundle from the server REPLACES the
    summary fields and prunes ``recent_turns`` whose timestamp falls at
    or before the new compaction's ``end_timestamp``.
  * ``record_message`` / ``send_message`` APPENDS a raw turn to
    ``recent_turns`` so the SDK can serve the live tail without
    round-tripping the server between compactions.
  * Eviction is LRU (max conversations) + age-since-last-activity. There
    is no per-entry TTL — ST stays valid until the next compaction
    arrives or the conversation goes idle for ``max_age``.

The class is thread-safe (RLock) so callers can mix sync and async
contexts without external coordination.

See ``docs/internal/sdk_authoritative_short_term_context_plan.md`` in
the cloud repo for the full design.
"""
from __future__ import annotations

import logging
import threading
from collections import OrderedDict
from dataclasses import dataclass, field
from datetime import datetime, timedelta, timezone
from typing import Any, Dict, List, Optional


logger = logging.getLogger("synap.sdk.cache.short_term_store")

DEFAULT_MAX_CONVERSATIONS = 100
DEFAULT_MAX_AGE = timedelta(hours=12)


@dataclass
class CachedShortTermContext:
    """Per-conversation cached short-term context.

    Mirrors ``ConversationContextProto`` plus a few SDK-side bookkeeping
    fields (``last_activity_at`` for LRU/age eviction).
    """

    conversation_id: str
    summary: Optional[str] = None
    factual_paragraph: Optional[str] = None
    conversational_paragraph: Optional[str] = None
    current_state: Dict[str, Any] = field(default_factory=dict)
    key_extractions: Dict[str, Any] = field(default_factory=dict)
    compaction_id: Optional[str] = None
    compacted_at: Optional[datetime] = None
    end_timestamp: Optional[datetime] = None
    recent_turns: List[Dict[str, Any]] = field(default_factory=list)
    last_activity_at: datetime = field(
        default_factory=lambda: datetime.now(timezone.utc)
    )

    def has_compaction(self) -> bool:
        return self.compaction_id is not None

    def to_dict(self) -> Dict[str, Any]:
        """Serializable view, suitable for emitting through telemetry or
        for handing to a local formatter."""
        return {
            "conversation_id": self.conversation_id,
            "summary": self.summary,
            "factual_paragraph": self.factual_paragraph,
            "conversational_paragraph": self.conversational_paragraph,
            "current_state": dict(self.current_state),
            "key_extractions": dict(self.key_extractions),
            "compaction_id": self.compaction_id,
            "compacted_at": self.compacted_at.isoformat() if self.compacted_at else None,
            "end_timestamp": self.end_timestamp.isoformat() if self.end_timestamp else None,
            "recent_turns": list(self.recent_turns),
        }


class ShortTermContextStore:
    """LRU cache of per-conversation ST blocks. Thread-safe."""

    def __init__(
        self,
        max_conversations: int = DEFAULT_MAX_CONVERSATIONS,
        max_age: timedelta = DEFAULT_MAX_AGE,
    ) -> None:
        self._lock = threading.RLock()
        self._cache: "OrderedDict[str, CachedShortTermContext]" = OrderedDict()
        self._max = max_conversations
        self._max_age = max_age

    # ------------------------------------------------------------------
    # Reads
    # ------------------------------------------------------------------

    def get(self, conversation_id: str) -> Optional[CachedShortTermContext]:
        """Return the cached entry or ``None``. Age-evicts on read."""
        if not conversation_id:
            return None
        with self._lock:
            entry = self._cache.get(conversation_id)
            if entry is None:
                return None
            if datetime.now(timezone.utc) - entry.last_activity_at > self._max_age:
                self._cache.pop(conversation_id, None)
                return None
            self._cache.move_to_end(conversation_id)
            return entry

    def has(self, conversation_id: str) -> bool:
        return self.get(conversation_id) is not None

    def size(self) -> int:
        with self._lock:
            return len(self._cache)

    # ------------------------------------------------------------------
    # Writes
    # ------------------------------------------------------------------

    def append_turn(
        self,
        conversation_id: str,
        role: str,
        content: str,
        timestamp: Optional[datetime] = None,
    ) -> None:
        """Append a raw turn to ``recent_turns``.

        Called from ``record_message`` / ``send_message``. If no cache
        entry exists yet (first turn before any compaction has arrived),
        a fresh entry is created with an empty summary; the next
        ``compaction_update`` bundle will fill in the summary fields and
        prune turns covered by it.
        """
        if not conversation_id:
            return
        ts = timestamp or datetime.now(timezone.utc)
        with self._lock:
            entry = self._cache.get(conversation_id)
            if entry is None:
                entry = CachedShortTermContext(conversation_id=conversation_id)
                self._cache[conversation_id] = entry
            entry.recent_turns.append(
                {
                    "role": role,
                    "content": content,
                    "timestamp": ts.isoformat(),
                }
            )
            entry.last_activity_at = ts if ts.tzinfo else ts.replace(tzinfo=timezone.utc)
            self._cache.move_to_end(conversation_id)
            self._evict_if_needed()

    def apply_compaction(self, bundle: Dict[str, Any]) -> None:
        """Apply a ``compaction_update`` bundle.

        Replaces the cached summary fields with the bundle's
        ``conversation_context`` payload and prunes locally-cached
        ``recent_turns`` whose timestamp falls at or before the
        compaction's ``end_timestamp`` (turns that arrived during the
        compaction window are preserved by design).

        Bundle shape matches the dict produced by ``_proto_to_bundle_dict``
        in ``transport/grpc_client.py``.
        """
        if not isinstance(bundle, dict):
            return
        cc = bundle.get("conversation_context") or {}
        conv_id = (
            cc.get("conversation_id")
            or bundle.get("_anticipation_conversation_id")
            or bundle.get("conversation_id")
        )
        if not conv_id:
            logger.debug("apply_compaction: bundle has no conversation_id, ignoring")
            return

        end_ts = _parse_iso(cc.get("end_timestamp") or cc.get("compacted_at"))
        compacted_at = _parse_iso(cc.get("compacted_at"))

        with self._lock:
            entry = self._cache.get(conv_id) or CachedShortTermContext(
                conversation_id=conv_id
            )
            entry.summary = cc.get("summary") or entry.summary
            entry.factual_paragraph = (
                cc.get("factual_paragraph") or entry.factual_paragraph
            )
            entry.conversational_paragraph = (
                cc.get("conversational_paragraph") or entry.conversational_paragraph
            )
            entry.current_state = dict(cc.get("current_state") or {})
            entry.key_extractions = dict(cc.get("key_extractions") or {})
            entry.compaction_id = cc.get("compaction_id") or entry.compaction_id
            entry.compacted_at = compacted_at or entry.compacted_at
            entry.end_timestamp = end_ts or entry.end_timestamp

            if end_ts is not None and entry.recent_turns:
                entry.recent_turns = [
                    t
                    for t in entry.recent_turns
                    if _is_after(t.get("timestamp"), end_ts)
                ]
            entry.last_activity_at = datetime.now(timezone.utc)
            self._cache[conv_id] = entry
            self._cache.move_to_end(conv_id)
            self._evict_if_needed()

    def invalidate(self, conversation_id: str) -> None:
        with self._lock:
            self._cache.pop(conversation_id, None)

    def clear(self) -> None:
        with self._lock:
            self._cache.clear()

    # ------------------------------------------------------------------
    # Internal
    # ------------------------------------------------------------------

    def _evict_if_needed(self) -> None:
        while len(self._cache) > self._max:
            evicted_id, _ = self._cache.popitem(last=False)
            logger.debug("Evicted ST cache entry (LRU) conversation_id=%s", evicted_id)


# ----------------------------------------------------------------------
# Helpers
# ----------------------------------------------------------------------


def _parse_iso(value: Any) -> Optional[datetime]:
    """Parse an ISO-8601 string into an aware datetime, or return None."""
    if value is None or value == "":
        return None
    if isinstance(value, datetime):
        return value if value.tzinfo else value.replace(tzinfo=timezone.utc)
    try:
        # ``fromisoformat`` accepts both ``…+00:00`` and naive forms; also
        # handle the ``Z`` suffix that the server occasionally emits.
        if isinstance(value, str) and value.endswith("Z"):
            value = value[:-1] + "+00:00"
        dt = datetime.fromisoformat(value)
        if dt.tzinfo is None:
            dt = dt.replace(tzinfo=timezone.utc)
        return dt
    except (TypeError, ValueError):
        logger.debug("Could not parse ISO timestamp: %r", value)
        return None


def _is_after(turn_timestamp: Any, cutoff: datetime) -> bool:
    """True if the turn's timestamp is strictly after the cutoff."""
    parsed = _parse_iso(turn_timestamp)
    if parsed is None:
        # Defensive: when we can't parse the timestamp, keep the turn
        # rather than silently drop it.
        return True
    return parsed > cutoff

File: packages/sdks/maximem-synap-js/src/cache/short-term-store.ts
/**
 * Per-conversation short-term context, held locally.
 *
 * Mirrors Python's `cache/short_term_store.py`. This is what makes reads see
 * your own writes: a turn recorded a moment ago is buffered here and spliced
 * onto the next fetch, so a fast follow-up does not retrieve context that omits
 * what the user just said. Python's own comment on the overlay puts it plainly:
 * the pre-compaction window "is exactly when the playground's fast follow-ups
 * happen."
 *
 * LRU by conversation, with an absolute age cap. Everything here is per process
 * and holds only what this process wrote, so nothing is authoritative: the
 * server's view wins for the summary fields, and this only ever contributes the
 * verbatim tail.
 */

const DEFAULT_MAX_CONVERSATIONS = 100;
/** 12 hours, as in Python. */
const DEFAULT_MAX_AGE_MS = 12 * 60 * 60 * 1000;

export interface ShortTermTurn {
  role: string;
  content: string;
  /** ISO-8601. */
  timestamp: string;
}

export interface ShortTermEntry {
  conversationId: string;
  summary: string | null;
  factualParagraph: string | null;
  conversationalParagraph: string | null;
  currentState: Record<string, unknown>;
  keyExtractions: Record<string, unknown>;
  compactionId: string | null;
  compactedAt: string | null;
  endTimestamp: string | null;
  recentTurns: ShortTermTurn[];
  lastActivityAt: number;
}

export interface ShortTermStoreOptions {
  maxConversations?: number;
  maxAgeMs?: number;
  /** Injected in tests. Wall clock, because entries are compared to ISO stamps. */
  now?: () => number;
}

/** Parse an ISO-8601 stamp to epoch millis, or null. */
export function parseIso(value: unknown): number | null {
  if (value === null || value === undefined || value === '') return null;
  if (value instanceof Date) return value.getTime();
  if (typeof value !== 'string') return null;
  const parsed = Date.parse(value);
  return Number.isNaN(parsed) ? null : parsed;
}

/**
 * True when a turn is strictly after the cutoff.
 *
 * An unparseable timestamp KEEPS the turn. Dropping it would silently lose a
 * message, which is worse than briefly showing one the server has already
 * folded into its summary.
 */
function isAfter(turnTimestamp: unknown, cutoffMs: number): boolean {
  const parsed = parseIso(turnTimestamp);
  if (parsed === null) return true;
  return parsed > cutoffMs;
}

export class ShortTermStore {
  // Insertion order IS the LRU order: re-inserting on touch moves an entry to
  // the end, so the first key Map yields is the least recently used.
  readonly #entries = new Map<string, ShortTermEntry>();
  readonly #maxConversations: number;
  readonly #maxAgeMs: number;
  readonly #now: () => number;

  constructor(options: ShortTermStoreOptions = {}) {
    this.#maxConversations = options.maxConversations ?? DEFAULT_MAX_CONVERSATIONS;
    this.#maxAgeMs = options.maxAgeMs ?? DEFAULT_MAX_AGE_MS;
    this.#now = options.now ?? (() => Date.now());
  }

  /** The entry, or null. Age-evicts on read, as Python does. */
  get(conversationId: string): ShortTermEntry | null {
    if (!conversationId) return null;
    const entry = this.#entries.get(conversationId);
    if (entry === undefined) return null;
    if (this.#now() - entry.lastActivityAt > this.#maxAgeMs) {
      this.#entries.delete(conversationId);
      return null;
    }
    this.#touch(conversationId, entry);
    return entry;
  }

  has(conversationId: string): boolean {
    return this.get(conversationId) !== null;
  }

  get size(): number {
    return this.#entries.size;
  }

  /**
   * Append a raw turn.
   *
   * Called from `record_message` and the stream's `send_message`. When no entry
   * exists yet (the first turn, before any compaction has arrived) a fresh one
   * is created with an empty summary; the next `compaction_update` fills the
   * summary fields in and prunes the turns it covers.
   */
  appendTurn(conversationId: string, role: string, content: string, timestamp?: string): void {
    if (!conversationId) return;
    const stamp = timestamp ?? new Date(this.#now()).toISOString();
    const entry = this.#entries.get(conversationId) ?? this.#blank(conversationId);
    entry.recentTurns.push({ role, content, timestamp: stamp });
    entry.lastActivityAt = parseIso(stamp) ?? this.#now();
    this.#touch(conversationId, entry);
    this.#evict();
  }

  /**
   * Apply a `compaction_update` bundle.
   *
   * Replaces the summary fields from the bundle's `conversation_context` and
   * prunes locally buffered turns at or before its `end_timestamp`. Turns that
   * arrived DURING the compaction window are kept by design: the server has not
   * seen them yet, which is the whole reason the local tail exists.
   */
  applyCompaction(bundle: Record<string, unknown>): void {
    if (typeof bundle !== 'object' || bundle === null) return;
    const cc = (bundle['conversation_context'] ?? {}) as Record<string, unknown>;
    const conversationId = String(
      cc['conversation_id'] ??
        bundle['anticipation_conversation_id'] ??
        bundle['conversation_id'] ??
        '',
    );
    if (!conversationId) return;

    const endTs = parseIso(cc['end_timestamp'] ?? cc['compacted_at']);
    const entry = this.#entries.get(conversationId) ?? this.#blank(conversationId);

    // `x || existing` throughout, matching Python: a blank field from the
    // server must not erase what is already held.
    entry.summary = (cc['summary'] as string) || entry.summary;
    entry.factualParagraph = (cc['factual_paragraph'] as string) || entry.factualParagraph;
    entry.conversationalParagraph =
      (cc['conversational_paragraph'] as string) || entry.conversationalParagraph;
    entry.currentState = { ...((cc['current_state'] as Record<string, unknown>) ?? {}) };
    entry.keyExtractions = { ...((cc['key_extractions'] as Record<string, unknown>) ?? {}) };
    entry.compactionId = (cc['compaction_id'] as string) || entry.compactionId;
    entry.compactedAt = (cc['compacted_at'] as string) || entry.compactedAt;
    entry.endTimestamp = (cc['end_timestamp'] as string) || entry.endTimestamp;

    if (endTs !== null && entry.recentTurns.length > 0) {
      entry.recentTurns = entry.recentTurns.filter((t) => isAfter(t.timestamp, endTs));
    }
    entry.lastActivityAt = this.#now();
    this.#touch(conversationId, entry);
    this.#evict();
  }

  invalidate(conversationId: string): void {
    this.#entries.delete(conversationId);
  }

  clear(): void {
    this.#entries.clear();
  }

  #blank(conversationId: string): ShortTermEntry {
    return {
      conversationId,
      summary: null,
      factualParagraph: null,
      conversationalParagraph: null,
      currentState: {},
      keyExtractions: {},
      compactionId: null,
      compactedAt: null,
      endTimestamp: null,
      recentTurns: [],
      lastActivityAt: this.#now(),
    };
  }

  /** Re-insert so this key becomes the most recently used. */
  #touch(conversationId: string, entry: ShortTermEntry): void {
    this.#entries.delete(conversationId);
    this.#entries.set(conversationId, entry);
  }

  #evict(): void {
    while (this.#entries.size > this.#maxConversations) {
      const oldest = this.#entries.keys().next();
      if (oldest.done === true) break;
      this.#entries.delete(oldest.value);
    }
  }
}
